### PR TITLE
Added HTTP file piped action for pdf support

### PR DESCRIPTION
**Pull request by sandeepanbaidya**

### CODE REVIEW

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        this.client = client;
+        this.handleBarsService = handleBarsService;
+    }
+
+    @Override
+    protected JsonNode execute(HttpFilePipedActionTemplate httpFilePipedActionTemplate, Workflow workflow) {
+        log.debug("Http File Piped Action triggered with Template: {} and Workflow: {}",
+                httpFilePipedActionTemplate, workflow);
+
+        String responseTranslator = httpFilePipedActionTemplate.getResponseTranslator();
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:55]: use val

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        HttpActionData httpActionData = transformSourcePayload(workflow, httpFilePipedActionTemplate);
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:56]: use val

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        log.debug("Action call data: {}", httpActionData);
+
+        if (httpFilePipedActionTemplate.isNoop()) {
+            log.warn("Returning noop response as NoOp is configured in template");
+            return httpFilePipedActionTemplate.getNoopResponse();
+        }
+        byte[] file = (byte[]) fetchFile(httpActionData);
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:63]: use val

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+        httpActionData = transformTargetPayload(workflow, httpFilePipedActionTemplate, file);
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:65]: val with new httpActionData2

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+        JsonNode httpResponse = handle(httpActionData);
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:66]: val

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        if (httpResponse != null && !Strings.isNullOrEmpty(responseTranslator)) {
+            return toJsonNode(handleBarsService.transform(responseTranslator, httpResponse));
+        }
+        return null;
+    }
+
+    private Object fetchFile(HttpActionData actionData) {
+        try(Response httpResponse = executeRequest(actionData) ) {
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:74]: val if works here

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    private JsonNode handle(HttpActionData actionData) {
+        try(Response httpResponse = executeRequest(actionData) ) {
+            val responseBodyStr = HttpUtil.body(httpResponse);
+            if (Strings.isNullOrEmpty(responseBodyStr)) {
+                return NullNode.getInstance();
+            }
+            log.debug("HTTP Response: {}", responseBodyStr);
+            List<String> contentType = Arrays.stream(
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:88]: val

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+                    .put("payload", responseBodyStr);
+        } catch (final Exception e) {
+            throw StatesmanError.propagate(e);
+        }
+    }
+
+    @SneakyThrows
+    private Response executeRequest(HttpActionData actionData) {
+        return actionData.getMethod().visit(new HttpMethod.MethodTypeVisitor<Response>() {
+            private final Map<String, String> headers = actionData.getHeaders();
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:107]: val

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+            private final String url = actionData.getUrl();
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:108]: val

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+            @Override
+            public Response visitPost() throws Exception {
+                log.info("HTTP_ACTION POST Call url:{}", url);
+                val payload = actionData.getPayload();
+                Response response = client.get().post(url, payload, headers);
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:114]: val

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+                            actionData, HttpUtil.body(response));
+                    throw new StatesmanError();
+                }
+                return response;
+            }
+
+            @Override
+            public Response visitGet() throws Exception {
+                log.info("HTTP_ACTION GET Call url:{}", url);
+                Response response = null;
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:126]: merge 126 and 127 with val

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+                return response;
+            }
+
+            @Override
+            public Response visitForm() throws Exception {
+                log.info("HTTP_ACTION FORM Call url:{}", url);
+                val payload = actionData.getPayload();
+                val file = actionData.getFile();
+                val fileMeta = actionData.getFileMeta();
+                Response response = client.get().form(url, file, payload, fileMeta, headers);
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:142]: val

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+        try {
+            return mapper.readTree(responseBodyStr);
+        } catch (Exception e) {
+            log.error("Error while converting to json:" + responseBodyStr, e);
+            return null;
+        }
+    }
+
+    private HttpActionData transformTargetPayload(Workflow workflow, HttpFilePipedActionTemplate httpFilePipedActionTemplate, byte[] file) {
+        JsonNode jsonNode = mapper.valueToTree(workflow);
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:169]: val

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+                    .file(file)
+                    .build();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    private HttpActionData transformSourcePayload(Workflow workflow, HttpFilePipedActionTemplate httpFilePipedActionTemplate) {
+        JsonNode jsonNode = mapper.valueToTree(workflow);
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:187]: val

File: statesman-engine/src/main/java/io/appform/statesman/engine/http/HttpClient.java [hunk 15]
```diff
@@ -54,4 +59,30 @@ public Response get(final String url,
         final Request request = getBuilder.build();
         return client.newCall(request).execute();
     }
+
+    public Response form(final String url, final byte[] file,
+                              final Map<String, String> payload,
+                              final Map<String, String> fileMeta,
+                              final Map<String, String> headers) throws IOException {
+        final HttpUrl httpUrl = HttpUrl.get(url);
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/http/HttpClient.java:67]: val

File: statesman-engine/src/main/java/io/appform/statesman/engine/http/HttpClient.java [hunk 16]
```diff
@@ -54,4 +59,30 @@ public Response get(final String url,
...
+
+        if(payload != null) {
+            payload.forEach(formBuilder::addFormDataPart);
+        }
+        if(fileMeta != null) {
+            for(String key : fileMeta.keySet())
+                formBuilder.addFormDataPart(key, fileMeta.get(key),
+                        RequestBody.create(APPLICATION_PDF, file));
+        }
+        Request.Builder postBuilder = new Request.Builder()
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/http/HttpClient.java:79]: val

File: statesman-engine/src/main/java/io/appform/statesman/engine/http/HttpClient.java [hunk 17]
```diff
@@ -54,4 +59,30 @@ public Response get(final String url,
...
+                .url(httpUrl)
+                .post(formBuilder.build());
+        if (headers != null) {
+            headers.forEach(postBuilder::addHeader);
+        }
+        final Request request = postBuilder.build();
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/http/HttpClient.java:85]: val

File: statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+    private Object fetchFile(HttpActionData actionData) {
+        try(Response httpResponse = executeRequest(actionData) ) {
+            return Objects.requireNonNull(httpResponse.body()).bytes();
+        } catch (final Exception e) {
+            throw StatesmanError.propagate(e);
+        }
+    }
+
+    private JsonNode handle(HttpActionData actionData) {
+        try(Response httpResponse = executeRequest(actionData) ) {
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/action/impl/HttpFilePipedAction.java:82]: val if works

File: statesman-engine/src/main/java/io/appform/statesman/engine/http/HttpClient.java [hunk 19]
```diff
@@ -54,4 +59,30 @@ public Response get(final String url,
         final Request request = getBuilder.build();
         return client.newCall(request).execute();
     }
+
+    public Response form(final String url, final byte[] file,
+                              final Map<String, String> payload,
+                              final Map<String, String> fileMeta,
+                              final Map<String, String> headers) throws IOException {
+        final HttpUrl httpUrl = HttpUrl.get(url);
+        MultipartBody.Builder formBuilder = new MultipartBody.Builder()
```
Comment (ravidesai47) [inline statesman-engine/src/main/java/io/appform/statesman/engine/http/HttpClient.java:68]: val

**Review by santanusinha — APPROVED** [review]

**Pull request merged**